### PR TITLE
[SR-4453] Provide aggregate operators for pipeline engine

### PR DESCRIPTION
**Pull request by liuyehcf**
Provide aggregate operators for pipeline engine

- AggregateBlockingSinkOperator
- AggregateBlockingSourceOperator
- AggregateStreamingSinkOperator
- AggregateStreamingSourceOperator

Roadmap

- [x] stage1: support AggregateBlockingOperator and AggregateStreamingOperator
- [x] stage2: split AggregateBlockingOperator/AggregateStreamingOperator into sink-source operator pair(this pull request)
- [ ] stage3: support AggregateDistinctBlockingSinkOperator and AggregateDistinctBlockingSourceOperator
- [ ] stage4: support AggregateDistinctStreamingSinkOperator and AggregateDistinctStreamingSourceOperator
- [ ] stage5: use LocalExchagneOperator to increase the degree of parallelism of aggregation

### CODE REVIEW

**Review by Pslydhh — DISMISSED** [review]

File: be/src/exec/pipeline/aggregate/aggregate_blocking_sink_operator.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        // for aggregate no group by, if _num_input_rows is 0,
+        // In update phase, we directly return empty chunk.
+        // In merge phase, we will handle it.
+        if (_aggregator->num_input_rows() == 0 && !_aggregator->is_needs_finalize()) {
+            _aggregator->set_ht_done();
+        }
+    }
+    COUNTER_SET(_aggregator->input_row_count(), _aggregator->num_input_rows());
+
+    _aggregator->sink_complete();
```
Comment (satanson) [inline be/src/exec/pipeline/aggregate/aggregate_blocking_sink_operator.cpp:48]: what's the difference between _is_sink_complete and _is_ht_done?

File: be/src/exec/pipeline/aggregate/aggregate_blocking_source_operator.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021 StarRocks Limited.
+
+#include "aggregate_blocking_source_operator.h"
+
+namespace starrocks::pipeline {
+
+bool AggregateBlockingSourceOperator::has_output() const {
```
Comment (satanson) [inline be/src/exec/pipeline/aggregate/aggregate_blocking_source_operator.cpp:7]: is_ht_done returning false means hash table's building is unfinished, so chunk is cannot pulled out.

File: be/src/exec/pipeline/aggregate/aggregate_blocking_source_operator.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    return _aggregator->is_sink_complete() && !_aggregator->is_ht_done();
+}
+
+bool AggregateBlockingSourceOperator::is_finished() const {
+    return _aggregator->is_sink_complete() && _aggregator->is_ht_done();
```
Comment (satanson) [inline be/src/exec/pipeline/aggregate/aggregate_blocking_source_operator.cpp:12]: seems wrong

**Review by satanson — APPROVED** [review]

**Review by satanson — CHANGES_REQUESTED** [review]
commented

File: be/src/exec/pipeline/aggregate/aggregate_blocking_sink_operator.cpp (hunk 1 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/aggregate/aggregate_blocking_sink_operator.cpp:48]: sink_complete means: no further chunk will push into sink operator

File: be/src/exec/pipeline/aggregate/aggregate_streaming_sink_operator.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    bool is_finished() const override;
+    void finish(RuntimeState* state) override;
+
+    Status prepare(RuntimeState* state) override;
+
+    StatusOr<vectorized::ChunkPtr> pull_chunk(RuntimeState* state) override;
+    Status push_chunk(RuntimeState* state, const vectorized::ChunkPtr& chunk) override;
+
+private:
+    Status _push_chunk_by_force_streaming();
```
Comment (kangkaisen) [inline be/src/exec/pipeline/aggregate/aggregate_streaming_sink_operator.h:26]: Would better add some comments for these three methods ?

File: be/src/exec/pipeline/aggregate/aggregate_blocking_sink_operator.cpp (hunk 1 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/aggregate/aggregate_blocking_sink_operator.cpp:48]: adjust is_ht_done to is_ht_eos

File: be/src/exec/pipeline/aggregate/aggregate_blocking_source_operator.cpp (hunk 2 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/aggregate/aggregate_blocking_source_operator.cpp:7]: adjust is_ht_done to is_ht_eos。true means all data in the hash table has been taken out

File: be/src/exec/pipeline/aggregate/aggregate_streaming_sink_operator.h (hunk 4 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/aggregate/aggregate_streaming_sink_operator.h:27]: fixed

**Review by kangkaisen — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**